### ISSUE TITLE
Add support for "missing" to all bucket aggregations

### ISSUE THREAD

**Issue by roytmana** [elastic/elasticsearch#5324]
NEED: In many (if not majority cases) when present users with business analytics, the user would want to see numbers for complete data set. No matter how you aggregate it should present the same data with the same number of documents. Inability to handle "missing" values exclude those from analysis making analyzed data set incomplete and grand totals dependent on which field(s) the aggregation is done. It is impossible to explain to the users why the lower level totals do not add up to the upper level ones!  

WORKAROUND: Currently field based bucket aggregations (term, range etc) have no way to aggregate missing values. The only way is to use missing aggregation on the same level and the same field as the term aggregation itself. It is easy enough when dealing with one level aggregations but if you have 2-3 level aggregation number of "missing" aggregations (and complete lower level aggregation to be repeated in them) mushrooms very quickly to the point that the query is huge, convoluted and not debuggable. It may affect performance  as well. Also fetched date needs to be heavily post-processed to extract multiple levels aggregation buckets from under various "missing" elements and put them inline with the regular aggregation values. Below please see a simple query to do 2 level aggregation with just one sum metrics

PROPOSAL: I would suggest that any aggregation operating on a field should have a missing option. If missing config is specified, aggregation should accumulate missing values under that value and honor any nested aggregations within. It should never assume any value like 0 or _missing since it may clash with actual keys. If it is not specified the aggregation should skip missing values as it does now.

This approach makes it entirely compatible with existing logic and give developers complete control over whether to aggregate missing and under what key. In cases when it is not needed (and not specified) there will be no performance overhead. But when it needed it will work faster as we would not need to do missing aggregation and aggregations under it separately (same goes for "other" aggregation)

To be honest, I would love to see the same handling for "other" - documents that have not been included in aggregation due to the aggregation size constraints. Again the same rationale - ability to slice complete data set regardless of aggregation structure. It is just as needed as "missing" and just as troublesome to calculate but 
I could understand if you did not add it as it may be not compatible with your algorithms but  PLEASE PLEASE add "missing" handling at least

```
{
      "total": {
        "sum": {
          "field": "money.totals.obligationTotal"
        }
      },
      "missing": {
        "missing": {
          "field": "division"
        },
        "aggs": {
          "total": {
            "sum": {
              "field": "money.totals.obligationTotal"
            }
          },
          "missing": {
            "missing": {
              "field": "fy"
            }
          },
          "group": {
            "terms": {
              "field": "fy",
              "order": { "_term": "asc" }
            },
            "aggs": {
              "total": {
                "sum": {
                  "field": "money.totals.obligationTotal"
                }
              }
            }
          }
        }
      },
      "group": {
        "terms": {
          "field": "division",
          "order": { "_term": "asc" },
          size:100
        },
        "aggs": {
          "total": {
            "sum": {
              "field": "money.totals.obligationTotal"
            }
          },
          "missing": {
            "missing": {
              "field": "fy"
            },
            "aggs": {
              "total": {
                "sum": {
                  "field": "money.totals.obligationTotal"
                }
              }
            }
          },
          "group": {
            "terms": {
              "field": "fy",
              "order": { "_term": "asc" }
            },
            "aggs": {
              "total": {
                "sum": {
                  "field": "money.totals.obligationTotal"
                }
              }
            }
          }
        }
      }
    }
```

cc @uboness, @jpountz

**Comment by rightaway** [comment]
+1

**Comment by hadrienk** [comment]
+1

**Comment by binarybarry** [comment]
+1.  Would love to have this feature restored soon.  At the very least, simply having an optional 'other' bucket with just a count would be quite useful.

**Comment by ianneub** [comment]
:+1: We make use of a lot of pie charts that need to show "other" fields. Now that Kibana 4 is out, this is even more important for us.

**Comment by sqpdln** [comment]
+1

**Comment by bowlby** [comment]
+1

**Comment by Asimov4** [comment]
+1

**Comment by cameronkerrnz** [comment]
+1. If my ELK stack can't help me to solidly point the finger at something --- because signals get weakened if I can't tell how significant the top-N things are --- then it becomes much harder to use Kibana (4) effectively. It effectively makes me want to keep Kibana 3 around.

**Comment by Nickology** [comment]
+1

**Comment by oliviervaussy** [comment]
+1 for the "other_bucket"

**Comment by roytmana** [comment]
looks like a lot of people have the same need as I do (bucketing on _missing and _other)
I posted this issue initially because it was a critical and pretty obvious need for my us case (which is a generic traditional analytics/data mart/data slicing and visualization system). Sadly, even though the issue has one of the highest ratings thanks to all your +1s :-))  after all these months it has not even got a planned version assigned to it... 
I personally can't get off the now deprecated facets without it

**Comment by jpountz** [comment]
> Sadly, even though the issue has one of the highest ratings thanks to all your +1s :-)) after all these months it has not even got a planned version assigned to it... 

The reason why this feature has not been implemented entirely is that it is very challenging. Your complaint feels like we are leaving users in a dead end but I don't think it's true. As a follow-up of this issue we added the ability to get the document count for other buckets (https://github.com/elastic/elasticsearch/pull/8213) and for documents that miss a value for the field, it is still possible to use the `missing` aggregation. I agree it might not be as user-friendly as getting a `missing` bucket but on the other hand there are very valid reasons to not have the same sub aggregations for documents that miss a value and we do not want to clutter the API with options for every possible use-case.

In addition, we are currently exploring new ways to work on top of aggregations in https://github.com/elastic/elasticsearch/issues/9876. It is not clear yet whether it will help on this issue but it will at least open new doors.

**Comment by roytmana** [comment]
@jpountz I appreciate the update and I realize that it is a complex matter but it is a very much needed for this rather common use case (whether for developer using ES or ELK users) 

I am not really complaining I am just making an observation that it has been a year since I submitted the issue and it seen fairly high interest from the user community and yet has not received a target version number thus it is quite possible that it will not get implemented (for what could be very valid technical or strategic reasons). If that's so I personally would like to know it so I concentrate on finding a solution. As you said, missing + total aggregations together with some transformation logic would allow to calculate _other and _missing but for complex scenarios of UI driven dynamic multilevel aggregations, such queries quickly become really huge and complicated and lot more expensive than even an un-optimized built-in solution.

So if the consensus of the ES  Team that implementation of _other and _missing buckets is not feasible, I would like to know it and start working on a solution that expands a concise "logical" query defining which agg should take _missing  and/or other into account into a large ES ("physical") one and then transform results to calculate and inject _other and _missing buckets on all levels of the result tree transparently. 

Thank you 
Alex

**Comment by kcberg** [comment]
Really need this too and it's the biggest barrier in moving to kibana 4 IMO. +1

**Comment by u238** [comment]
I think that without that functionality the "top N" feature is really not that useful.

+1

**Comment by aelagon** [comment]
+1

**Comment by kstachowiak-kcura** [comment]
+1

**Comment by jpountz** [comment]
Update: Now that most aggregations support a missing option (https://github.com/elastic/elasticsearch/pull/11042) and that the terms aggregation returns counts for the `other` bucket (https://github.com/elastic/elasticsearch/pull/8213), I think the only remaining question is whether we should add an `other` bucket.

While I'm pretty happy with the way we deal with missing, I'm concerned about adding an `other` bucket. One idea could be to make it optional, but I don't like adding options that feel unnecessary. Besides I have no idea whether the default should be true or false. On the other hand if we enforce it to be computed, I'm afraid of the multi-valued case because either we try to make it behave like a regular bucket and count documents only once which would be costly, or we just merge other buckets naively and then the multi-valued case would be wrong/confusing because some documents could be counted several times. This is the reason why the other count is currently labeled as `sum_other_doc_count` to make clear it is the sum of the doc counts of other buckets and not the document count that an `other` bucket would have.

**Comment by Kallin** [comment]
My main reason for +1 this issue was that I wanted an other bucket for filter aggs. I found it inconvenient to have to duplicate the entire filter agg (which can be very complex) within a parallel filter agg that has an enclosing 'not' filter to get the remainder bucket. Afaik that's still not taken care of, so no simple way to get the other remainder/other bucket for a filter agg.

**Comment by jpountz** [comment]
Indeed `filters` is a different story, and I think we could implement easily. I opened https://github.com/elastic/elasticsearch/issues/11289 to track it separately.

**Comment by roytmana** [comment]
I would say "other" is very useful as well particularly for in case people implement data mart like systems with elasticsearch. They tend to present complete dataset rollup to the user where "other" is needed to have the complete dataset representation. Also in traditional datamarts with start like data models they would avoid any multivalued relations instead modeling it as multiple fact "tables" so the issue with counting "other" multiple times would not be especially relevant to these use cases. And when people facet on multivalue field they would need to understand multiplicity issue just like they need to understand that sum of counts of its buckets is not the number of docs but values. 

I would like "other" to be a fully supported bucket enabled only if user explicetly specified it to be calculated by providing a key value for it. Ithink the case of single value aggs where it works as expected intuitively is so useful in itself as to justify this feature and the multivalued case would need to be understood and dealt with by each elastic user

Just my 2c

Jist

**Comment by anshaw** [comment]
+1

**Comment by elwerene** [comment]
+1

**Comment by dev-shubh** [comment]
+1

**Comment by ymost** [comment]
+1

**Comment by jpountz** [comment]
This is actually now implemented via #11042 and will be available in elasticsearch 2.0.

**Comment by jpountz** [comment]
Hmm, just remembered that in spite of the title, this issue was not only about missing but also about an other bucket. I consider the `missing` bucket issue solved, and the `other` bucket half-solved given that the `filters` aggregation now supports an `other` bucket and that the `terms` aggregation can return the sum of the doc counts for "other" buckets. If you would like to keep the discussion going about the `other` bucket, please open a new issue.

**Comment by funbrain** [comment]
hi there

let me ask one thing.

my scenario is that

let's say i have 3 requests to A system and i expected 3 responses from A system.
but in real i got 2 responses from A system.
one response was dropout.
probably, user close browser while A system was progressing....

eg raw msg:

> 2017-12-19 09:25:50,207 (null) 27 Info : ASystem : TESTPAGE :: elkko5zhjzyigt4it1wjmxpw :: Request : <UserRequest><version>9.0</version><RefID>764764000000881</RefID></UserRequest>
> 2017-12-19 09:25:50,207 (null) 27 Info : ASystem : TESTPAGE :: elkko5zhjzyigt4it1wjmxpw :: Response: <UserResponse><version>9.0</version><RefID>764764000000881</RefID></UserResponse>

> 2017-12-19 09:26:50,207 (null) 27 Info : ASystem : TESTPAGE :: elkko5zhjzyigt4it1wjmxpw :: Request : <UserRequest><version>9.0</version><RefID>764764000000882</RefID></UserRequest>

> 2017-12-19 09:27:50,207 (null) 27 Info : ASystem : TESTPAGE :: elkko5zhjzyigt4it1wjmxpw :: Request : <UserRequest><version>9.0</version><RefID>764764000000883</RefID></UserRequest>
> 2017-12-19 09:27:50,207 (null) 27 Info : ASystem : TESTPAGE :: elkko5zhjzyigt4it1wjmxpw :: Response: <UserResponse><version>9.0</version><RefID>764764000000883</RefID></UserResponse>

Above A system logs stream to logstash
Then logstash make some fields using grok patterns.

eg fields:

> log_timestamp: 2017-12-19 09:27:50,207
> log_level: Info
> log_type: REQUEST *if log msg have Request then log_type: REQUEST
> log_type: RESPONSE *if log msg have Response then log_type: RESPONSE
> ref_id: 764764000000883
> version: 9.0


In Request and Response logs, there is no response log found in second request?
My question is that how to query for missing response log count (dropout) in elasticsearch????

plz, advise to me. Thank you
